File: src/fwd-svc.py
#!/opt/homebrew/bin/python3
import re
import argparse
import os
import sys


def get_running_forwards():
    cmd = "ps -e -o command | grep port-forward | grep -v grep"
    # cmd = "/Users/dibokette/bin/port-forward_temp"
    stream = os.popen(cmd)
    return [s.strip() for s in stream.readlines()]


def get_pods(pod_filter):
    if not pod_filter or re.search(r'\s', pod_filter):
        cmd = "kubectl get pods | grep staging"
        # cmd = "/Users/dibokette/bin/kpods_temp | grep staging"
    else:
        cmd = f"kubectl get pods  | grep staging | grep {pod_filter}"
        # cmd = f"/Users/dibokette/bin/kpods_temp | grep staging | grep {pod_filter}"
    stream = os.popen(cmd)
    stripped = [s.strip() for s in stream.readlines()]
    return [re.sub(r'^(\S+).*$', r'\1', s) for s in stripped]


parser = argparse.ArgumentParser()
parser.add_argument("--grep", help="search pattern")
parser.add_argument("--ps", help="print the running local service forwarders", action="store_true")
parser.add_argument("--pods", help="print the running pods", action="store_true")
parser.add_argument("--lport", help="local port")
parser.add_argument("--pport", help="pod port")
args = parser.parse_args()

local_port = 8080
pod_port = 8080

if args.ps:
    output = get_running_forwards()
    for x in range(len(output)):
        print(output[x], end='')
    exit(0)

if args.pods:
    output = get_pods(args.grep)
    for x in range(len(output)):
        print(output[x], end='')
    exit(0)

if args.lport:
    local_port = int(args.lport)

if args.pport:
    pod_port = int(args.pport)

running = get_running_forwards()
pods = get_pods(args.grep)

if len(pods) == 0:
    print("Did not get any pods", file=sys.stderr)
    exit(1)

name_to_running_map = {re.sub(r'^\w+\s[\w-]+\s(\w+)-staging-.+$', r'\1', running[i]): running[i] for i in
                       range(0, len(running))}

name_to_pod_map = {re.sub(r'^([\w-]+?)-staging-.+$', r'\1', pods[i]): pods[i] for i in range(0, len(pods))}

# remove entries from name_to_pod_map that are in name_to_running_map
for key in name_to_running_map.keys():
    name_to_pod_map.pop(key, None)

podname_ordered = sorted(name_to_pod_map.keys())

for index in range(0, len(podname_ordered)):
    print(f"{index}: {podname_ordered[index]}")

ans = input("which one?")

if re.match(r'^quit|^q', ans, re.IGNORECASE):
    exit(0)

if not re.match(r'^\d+$', ans):
    print("Need a number", file=sys.stderr)
    exit(1)

chosen_idx = int(ans)

if chosen_idx > len(name_to_pod_map):
    print("Number is out of range", file=sys.stderr)
    exit(1)

chosen_pod = name_to_pod_map[podname_ordered[chosen_idx]]

kcmd = f'kubectl port-forward {chosen_pod} {local_port}:{pod_port} &'

print("Currently Running:")
print(running)
print()

ans = input(f'Run command: "{kcmd}"?')

if not re.match(r'^y', ans, re.IGNORECASE):
    print("Cancelling", file=sys.stderr)
    exit(0)

print(f'Running command: "{kcmd}"')
os.system(kcmd)

File: src/fwd-svc.pl
#!/usr/bin/perl
use strict;
use feature ":5.10";
use warnings FATAL => 'all';

use Getopt::Long;

my $promptForKill = 1;

sub killAllRunning() {
    chomp(my @lines = `ps -ef | grep port-forward | grep -v grep`);

    my @killPidList = ();
    foreach my $line (@lines){
        say $line;
        (my $pid = $line) =~ s/^\s*\d+\s*(\d+)\s.+$/$1/;
        if (defined $pid){
            push(@killPidList, $pid);
            say "pid to kill: $pid"
        }
    }

    if (@killPidList == 0){
        say "No PIDs to kill.";
        return;
    }

    my $input = "y";
    if ($promptForKill != 0){
        print "kill these[y/n]? ";
        $input = <STDIN>;
    }

    if ($input =~ /^y/i){
        foreach my $pid (@killPidList){
            system("kill", $pid);
        }
    }
}

sub getFilename() {
    my $DIR = $ENV{FWD_SVC_CONFIG_DIR} // "$ENV{HOME}/forwards";
    chomp(my @files = `ls -1 $DIR | grep '.csv\$'`);

    my $index = 0;
    foreach my $file (@files) {
        my $pos = ++$index;
        say "${pos}) $file";
        system("cat $DIR/$file | xargs -n 1 echo '    '")
    }

    print "which number? ";
    my $input = <STDIN>;

    if ($input !~ /^\d+$/) {
        die "Didn't get a number";
    }

    if ($input < 1 || $input > $index){
        die "invalid number";
    }

    return "$DIR/$files[$input - 1]";
}

#####################################################
# Main Script
#####################################################

my $filename;
chomp(my $kenv = `kubens -c`);
my $ps;
my $killall;
(my $exeName = $0) =~ s/^(?:.+\/)([^\/]+)$/$1/;
my $USAGE = "usage: $exeName --file <file> | $exeName --[ps|ls] | $exeName --killall";

GetOptions(
    "file=s" => \$filename,
    "killall" => \$killall,
    "ps" => \$ps,
    "ls" => \$ps,
) or die $USAGE;

if (defined $ps) {
    system("ps -e -o command | grep port-forward | grep -v grep");
    exit $?;
}

if (defined $killall) {
    killAllRunning();
    exit(0);
}

if (!defined $filename) {
    $filename = getFilename();
    if (!defined $filename) {
        say STDERR "filename is empty";
        die $USAGE;
    }
}

if ($kenv !~ /staging|sandbox|production/) {
   die "Invalid value for kenv"
}

if (! -r $filename){
    die "Not readable: $filename";
}

chomp(my @lines = `cat $filename`);

# File format is: pod:port

my %desiredPodToPortMap = ();
foreach my $line (@lines) {
    $line =~ s/^(.*?)#.*$/$1/;

    if ($line =~ /^\s*$/) {
        next;
    }

    my ($pod, $port) = split(/,/, $line, -1);
    # say "pod=$pod";
    # say "port=$port";
    $desiredPodToPortMap{$pod} = $port;
}

chomp(my @running = `ps -e -o command | grep port-forward | grep -v grep`);
# chomp(my @running = `/Users/dibokette/bin/port-forward_temp`);
my %runningPodToLocalPortMap = ();
foreach my $runEntry (@running){
    if ($runEntry =~ m/^.+port-forward (.+?)-$kenv\S+ (\d+).+$/) {
        $runningPodToLocalPortMap{$1} = $2;
    }
}

chomp(my @kpods = `kubectl get pods | cut -d ' ' -f 1`);
# chomp(my @kpods = `/Users/dibokette/bin/kpods_temp | grep $kenv | cut -d ' ' -f 1`);
$? != 0 and die "Could not run kubectl - need to run avp?";

my %podToFullnameMap = ();
foreach my $kpod (@kpods){
    (my $podname = $kpod) =~ s/(^.+?)-${kenv}-.+$/$1/;
    # $kpod =~ s/(^.+?)-${kenv}-.+$/$1/;
    $podToFullnameMap{$podname} = $kpod;
}

say "kpods - filtered on pods we're currently running or wanting to run (KPODS map):";
foreach my $pod (sort {$a cmp $b} keys %podToFullnameMap){
    if (exists $desiredPodToPortMap{$pod} || exists $runningPodToLocalPortMap{$pod}){
        say "$pod -- $podToFullnameMap{$pod}";
    }
}
say "";

say "Pods we are currently running and their local port (RUNNING map):";
foreach my $runningPod (sort {$a cmp $b} keys %runningPodToLocalPortMap){
    say "$runningPod -- $runningPodToLocalPortMap{$runningPod}";
}
say "";

say "Pods we want to have running in the end-state (DESIRED map):";
foreach my $desiredPod (sort {$a cmp $b} keys %desiredPodToPortMap){
    say "$desiredPod -- $desiredPodToPortMap{$desiredPod}";
}
say "";

say "Commands to run:";
my @diffCommandList = ();
my @allCommandList = ();
foreach my $desiredPod (sort {$a cmp $b} keys %desiredPodToPortMap){
    my $port = $desiredPodToPortMap{$desiredPod};
    my $remotePort = "8080";
    if ($port =~ m/^(\d+):(\d+)$/) {
        $port = $1;
        $remotePort = $2;
    }

    my $podFullname = $podToFullnameMap{$desiredPod};
    if (!defined $podFullname){
        die "Unable to find pod $desiredPod via kpods command"
    }

    my $cmd = "kubectl port-forward $podFullname $port:$remotePort &";
    say $cmd;
    push (@allCommandList, $cmd);

    if (exists $runningPodToLocalPortMap{$desiredPod} && $runningPodToLocalPortMap{$desiredPod} == $port) {
        # this pod is already running
        next;
    }

    # say $cmd;
    push (@diffCommandList, $cmd);
}
say "";

if ((@diffCommandList == 0) && ((keys %desiredPodToPortMap) == (keys %runningPodToLocalPortMap))) {
    say "The running commands seem to be the same as desired commands - so not doing anything.";
    say "If you want to re-run them anyways, do `$exeName --killall` and then run this command again.";
    exit(0);
}

# print "run these kubectl commands[y/n]? ";
print "kill all current running, and run these commands? [y/n]? ";
my $input = <STDIN>;

if ($input =~ /^y/i){
    # foreach my $cmd (@diffCommandList){
    #     system($cmd);
    # }
    $promptForKill = 0;
    killAllRunning();
    sleep(2);
    foreach my $cmd (@allCommandList){
        system($cmd);
    }
} else {
    say "Doing nothing";
    exit(0);
}
